prometheus/src/cluster_metrics.rs

```diff
@@ -25,18 +39,105 @@ fn get_vote_state(bank: &Bank, vote_pubkey: &Pubkey) -> Option<ValidatorVoteInfo
     let vote_state = vote_account.vote_state();
     let vote_state = vote_state.as_ref().unwrap_or(&default_vote_state);
 
+    let identity = vote_state.node_pubkey;
+
+    let validator_info: ValidatorInfo = match get_validator_info_accounts(bank) {
```

Comment:
> get_validator_info_accounts is very expensive, it builds a hash map of size O(num_validators_with_info), but for that it has to do a linear scan over O(num_config_accounts) accounts, and I suspect that without account indexing enabled (we don’t enable account indexing on our validator nodes, only on the RPCs), it has to do a linear scan over all Solana accounts to find the ones owned by the config program.

Yeah so the implementation was done to pretty much fit the current design and I opted for not introducing premature optimization. I have noticed that this is pretty expensive computation, tho given that this is only internal endpoint that is called every 5 sec it shouldn't have that huge of an impact. Also since all `Bank` data is already in memory we are not doing any IO etc.

I was thinking about things you mentioned as well - constructing the map once per bank, and only for the monitored vote accounts tho did not think the optimization is worth it (I do not know what size are we talking about in regards to `Bank`, couple of KB, MB, GB? so it is hard to judge). 

> I’m being more cautious here then we are in the Solido maintainer daemon, because there is more at stake.

Yeah if we want to be very cautious, then we can do the optimizations for sure.

Reply:
I don’t think it’s premature, if you try to run the `getProgramAccounts` RPC call against a node that doesn’t have account indexing for the config program, it times out. It’s difficult to test locally because even if you create 1000 vote accounts, there will not be any other accounts so it’s still pretty fast. But I think the production validator really has to walk many many more accounts. Now that we are here, we can look at how `get_program_accounts` is implemented.

Reply:
We discussed a bit with @enriquefynn, and concluded that it might be worth fetching this data just once on validator startup and storing it together with monitored vote accounts. While we risk having a stale name in metrics if validator changes their name I think it should be fine since it is only for convenience.

Reply:
Yeah exactly, I don’t expect name changes to be a problem, and if they are it’s also easy to fix; we can store the address of the config account and parse it during metrics fetch, instead of parsing it once at startup. Parsing ~10 config accounts to extract the name on every metrics request shouldn’t be a big deal.

The more difficult problem is when a validator creates a _second_ config account. Then we don’t know which one is the right one and we should ignore both. But if we already have the address of the first one and it remains there, we won’t notice a second one being created. I think that’s fine, it’s a very hypothetical thing, and the worst thing that happens is that we continue to report a name that we already considered valid in the past.

Reply:
> we can store the address of the config account and parse it during metrics fetch, instead of parsing it once at startup

Could you elaborate on that? Do you mean the reference to actual data structure or something else? If that is the case I am not sure if that is possible since from my understanding (might be very well wrong) the account data would reference specific slot so it would not change in case the data changes. Did some testing around that which seems to confirm this hypothesis, if you have some ideas how it could be done I would be happy to hear it.

Reply:
You could create a `struct` that lives in `RpcRequestMiddleware` and is initialized with a mapping from vote key to validator name.

Reply:
Yes, thanks, I get that part, was wondering about the comment about storing reference to Account data and parsing it during request - not sure if that is possible.

Reply:
Validator info is stored in config accounts, and the config account itself records which validator identity pubkey it stores info about. So to locate a validator’s info, we have to look at all config accounts. (In database terms, we would have a table with columns `config_accounts (addr pubkey, validator_identity pubkey)` without index on `validator_identity`, so we need to do a full table scan to look up the config for a validator.)

But after we look up a validator once, we know the config account that holds the info for that validator. We could put that in a `HashMap` that maps validator identity pubkey to config account pubkey. (In a sense, we manually build a partial index.) As Fynn says, we can store it in `RpcRequestMiddleware`. Then to look up validator info by identity pubkey, we would:

1. Check our hashmap if we know the address of the config account already.
2. If we don’t know it yet, do the expensive linear-time lookup, and put the address in the hashmap.
3. So now we do have the address of the config account, but if it came from the hashmap, in theory the config account could have changed and describe a different validator now. (Not sure if the config program allows that.) Or the config account could be deleted, or deleted and replaced with a fake account that holds valid data but is not owned by the config program. (Again, not sure if the config program allows this.)
4. So we have to verify that the config account is still good, that it is owned by the config program, that it describes the validator we were looking for, and that it is signed by that validator.
5. If the config account turns out to be invalid, delete the entry from the hash map, go back to 1.
6. If it is valid, we found the validator info, and usually we will have found it in constant time because we can skip step 2.

Then in practice we should probably implement it in such a way that step 2 can collect all unknown accounts in a single pass, instead of having to iterate all config accounts once per new identity account.

Does that make sense?